### PR TITLE
Pl/channelbases

### PR DESCRIPTION
**Pull request by plewandowska777**

### CODE REVIEW

File: src/matrixbases.jl [hunk 1]
```diff
@@ -55,6 +56,72 @@ function represent(basis::Type{T}, m::Matrix{<:Number}) where T<:AbstractMatrixB
...
+end 
+length(itr::ChannelBasisIterator) = itr.idim^2 * itr.odim^2 - itr.idim^2 + 1
+
+function represent(basis::ChannelBasis{T1}, Φ::AbstractQuantumOperation{T2}) where T1<:AbstractMatrix{<:Number} where T2<:AbstractMatrix{<:Number}
+    J = convert(DynamicalMatrix{T2}, Φ)
+    represent(basis, J.matrix) 
+end
+
+function combine(basis::ChannelBasis{T}, v::Vector{<:Number}) where T<:AbstractMatrix
+    m = sum(basis.iterator .* v)
```
Comment (lpawela) [inline src/matrixbases.jl]: this should call combine for AbstractMatrixBasis

File: src/matrixbases.jl [hunk 2]
```diff
@@ -55,6 +56,72 @@ function represent(basis::Type{T}, m::Matrix{<:Number}) where T<:AbstractMatrixB
...
+    function ChannelBasis{T}(idim::Integer, odim::Integer) where T<:AbstractMatrix{<:Number}
+        new(ChannelBasisIterator{T}(idim, odim)) 
+    end
+end
+channelbasis(T::Type{<:AbstractMatrix{<:Number}}, idim::Int, odim::Int=idim) = ChannelBasis{T}(idim, odim)
+
+channelbasis(idim::Int, odim::Int=idim) = channelbasis(Matrix{ComplexF64}, idim, odim)
+
+function iterate(itr::ChannelBasisIterator{T}, state=(1,1,1,1)) where T<:AbstractMatrix{<:Number}
+    idim = itr.idim
```
Comment (lpawela) [inline src/matrixbases.jl]: These assignments can be done in a single line

File: src/matrixbases.jl [hunk 3]
```diff
@@ -55,6 +56,72 @@ function represent(basis::Type{T}, m::Matrix{<:Number}) where T<:AbstractMatrixB
...
+    elseif d == idim && b == idim && c < odim
+        newstate = (a, c+1, 1, 1)
+    else  
+        newstate = (a+1, 1, 1, 1)
+    end 
+    return x, newstate
+end 
+length(itr::ChannelBasisIterator) = itr.idim^2 * itr.odim^2 - itr.idim^2 + 1
+
+function represent(basis::ChannelBasis{T1}, Φ::AbstractQuantumOperation{T2}) where T1<:AbstractMatrix{<:Number} where T2<:AbstractMatrix{<:Number}
```
Comment (lpawela) [inline src/matrixbases.jl]: The first argument should be AbstractChannelBasis

File: src/matrixbases.jl [hunk 4]
```diff
@@ -55,6 +56,72 @@ function represent(basis::Type{T}, m::Matrix{<:Number}) where T<:AbstractMatrixB
...
+function iterate(itr::ChannelBasisIterator{T}, state=(1,1,1,1)) where T<:AbstractMatrix{<:Number}
+    idim = itr.idim
+    odim = itr.odim
+    hitr = itr.hitr
+    (a, c, b, d) = state
+    (a == odim && c == odim && d == 2) && return nothing
+
+    Tn = eltype(T)
+    if a > c
+        x = (ketbra(T, a, c, odim) ⊗ ketbra(T, b, d, idim) + ketbra(T, c, a, odim) ⊗ ketbra(T, d, b, idim)) / sqrt(Tn(2))
```
Comment (lpawela) [inline src/matrixbases.jl]: This is almost the same as line 99. Maybe we can write them as one expression. The same holds for the cases in the HermiatianBasis

File: test/matrixbases.jl [hunk 5]
```diff
@@ -27,9 +35,41 @@ end
...
+    # @test eltype(vC) == eltype(C)
+    # @test length(vC) == prod(size(C))
+end
+
 @testset "hermitainbasis" begin
     @test hermitianbasis(Matrix{Float32}, 2) == HermitianBasisIterator{Matrix{Float32}}(2)
     @test hermitianbasis(2) == HermitianBasisIterator{Matrix{ComplexF64}}(2)
 end
 
+# @testset "channelbasis" begin
```
Comment (lpawela) [inline test/matrixbases.jl]: Why this is commented out?

File: test/matrixbases.jl [hunk 6]
```diff
@@ -27,9 +35,41 @@ end
     @test length(vC) == prod(size(C))
 end
 
+@testset "represent, combine" begin
```
Comment (lpawela) [inline test/matrixbases.jl]: the testset has the same name as the one on line 17

File: test/matrixbases.jl [hunk 7]
```diff
@@ -27,9 +35,41 @@ end
...
+    d1 = 2
```
Comment (lpawela) [inline test/matrixbases.jl]: Follow the convention, use idim and odim

File: test/matrixbases.jl [hunk 8]
```diff
@@ -27,9 +35,41 @@ end
...
+    @test A ≈ Ap.matrix
+    
+    A = reshape(collect(1:64), d1 * d2, d1 * d2) * reshape(collect(1:64), d1 * d2, d1 * d2)' + Matrix{Float64}(I, d1 * d2, d1 * d2)
+    B = Matrix{Float64}(I, d1, d1) ⊗ (ptrace(A, [d1, d2], 1))^(-1/2)
+    B = B * A * B'
+    vB = represent(ChannelBasis{Matrix{ComplexF64}}(d2, d1), B)
+    Bp = combine(ChannelBasis{Matrix{ComplexF64}}(d2, d1), vB)
+    @test B ≈ Bp.matrix
+
+    #vB = represent(ChannelBasis{Matrix{ComplexF32}}(d1,d2), B)
```
Comment (lpawela) [inline test/matrixbases.jl]: Why is this commented out?

**Comment by lpawela** [comment]
@plewandowska777 please resolve merge conflicts

File: src/matrixbases.jl [hunk 9]
```diff
@@ -55,6 +56,72 @@ function represent(basis::Type{T}, m::Matrix{<:Number}) where T<:AbstractMatrixB
...
-function combine(basis::T, v::Vector{<:Number}) where T<:AbstractMatrixBasis
+function combine(basis::AbstractMatrixBasis{T}, v::Vector{<:Number}) where T<:AbstractMatrix{<:Number}
     sum(basis.iterator .* v)
+end
+
+"""
+$(SIGNATURES)
+
+"""
+abstract type AbstractMatrixBasisIterator{T<:AbstractMatrix} end
```
Comment (pgawron) [inline src/matrixbases.jl]: Replication of `AbstractMatrixBasisIterator` from line 5.

File: src/matrixbases.jl [hunk 10]
```diff
@@ -55,6 +56,72 @@ function represent(basis::Type{T}, m::Matrix{<:Number}) where T<:AbstractMatrixB
...
+    idim::Int 
+    odim::Int
+    hitr::HermitianBasisIterator{T}
+    function ChannelBasisIterator{T}(idim::Int, odim::Int) where T<:AbstractMatrix{<:Number}
+        new(idim, odim, HermitianBasisIterator{T}(idim))
+    end
+end
+
+abstract type AbstractChannelBasis{T} <: AbstractMatrixBasis{T} end 
+struct ChannelBasis{T} <: AbstractMatrixBasis{T}
```
Comment (pgawron) [inline src/matrixbases.jl]: Should derive from `AbstractChannelBasis` not `AbstractMatrixBasis`.

File: src/matrixbases.jl (hunk 2 above)
Comment (pgawron) [reply src/matrixbases.jl]: I do not believe it is necessary.

File: src/matrixbases.jl [hunk 11]
```diff
@@ -55,6 +56,72 @@ function represent(basis::Type{T}, m::Matrix{<:Number}) where T<:AbstractMatrixB
...
+    elseif d == idim && b < idim
+        newstate = (a, c, b+1, 1)
+    elseif d == idim && b == idim && c < odim
+        newstate = (a, c+1, 1, 1)
+    else  
+        newstate = (a+1, 1, 1, 1)
+    end 
+    return x, newstate
+end 
+length(itr::ChannelBasisIterator) = itr.idim^2 * itr.odim^2 - itr.idim^2 + 1
```
Comment (pgawron) [inline src/matrixbases.jl]: Add one new line before new method definition.

File: src/matrixbases.jl [hunk 12]
```diff
@@ -55,6 +56,72 @@ function represent(basis::Type{T}, m::Matrix{<:Number}) where T<:AbstractMatrixB
...
+
+function represent(basis::ChannelBasis{T1}, Φ::AbstractQuantumOperation{T2}) where T1<:AbstractMatrix{<:Number} where T2<:AbstractMatrix{<:Number}
+    J = convert(DynamicalMatrix{T2}, Φ)
```
Comment (pgawron) [inline src/matrixbases.jl]: I believe we should coerce the numerical type with the type of the output, so: `DynamicalMatrix{T1}`.

File: test/matrixbases.jl [hunk 13]
```diff
@@ -6,6 +6,14 @@
     @test [tr(m[i]' * m[j]) for i=1:d, j=1:d] ≈ Matrix{Float64}(I, d, d)
 end
 
+@testset "ChannelBasisIterator" begin
+    d1 = 2
+    d2 = 2
```
Comment (pgawron) [inline test/matrixbases.jl]: By default do not test rectangular matrices. Add a test with d1=\= d2.

File: test/matrixbases.jl [hunk 14]
```diff
@@ -27,9 +35,41 @@ end
     @test length(vC) == prod(size(C))
 end
 
+@testset "represent, combine" begin
+    d1 = 2
+    d2 = 4
+    A = reshape(collect(1:16), d1 * d2, d1) * reshape(collect(1:16), d1 * d2, d1)'
+    B = Matrix{Float64}(I, d2, d2) ⊗ (ptrace(A, [d2, d1], 1))^(-1/2)
```
Comment (pgawron) [inline test/matrixbases.jl]: What will happen here of `B` is singular?

File: test/matrixbases.jl [hunk 15]
```diff
@@ -27,9 +35,31 @@ end
...
+    A = B * A * B'
+    vA = represent(ChannelBasis{Matrix{ComplexF64}}(idim, odim), A)
+    Ap = combine(ChannelBasis{Matrix{ComplexF64}}(idim, odim), vA)
+    @test A ≈ Ap.matrix
+    
+    A = reshape(collect(1:64), idim * odim, idim * odim) * reshape(collect(1:64), idim * odim, idim * odim)' + Matrix{Float64}(I, idim * odim, idim * odim)
+    B = Matrix{Float64}(I, idim, idim) ⊗ (ptrace(A, [idim, odim], 1))^(-1/2)
+    B = B * A * B'
+    vB = represent(ChannelBasis{Matrix{ComplexF64}}(odim, idim), B)
+    Bp = combine(ChannelBasis{Matrix{ComplexF64}}(odim, idim), vB)
```
Comment (lpawela) [inline test/matrixbases.jl]: Again, use `channelbasis`

File: test/matrixbases.jl [hunk 16]
```diff
@@ -27,9 +35,31 @@ end
...
+    @test eltype(vB) == Float64
+
+end
+
 @testset "hermitainbasis" begin
     @test hermitianbasis(Matrix{Float32}, 2) == HermitianBasisIterator{Matrix{Float32}}(2)
     @test hermitianbasis(2) == HermitianBasisIterator{Matrix{ComplexF64}}(2)
 end
 
+
```
Comment (lpawela) [inline test/matrixbases.jl]: Add the same tests for `channelbasis`

File: test/matrixbases.jl [hunk 17]
```diff
@@ -27,9 +35,31 @@ end
     @test length(vC) == prod(size(C))
 end
 
+@testset "representchannel, combinechannel" begin
+    (idim, odim) = (2,4)
+    A = reshape(collect(1:16), idim * odim, idim) * reshape(collect(1:16), idim * odim, idim)'
+    B = Matrix{Float64}(I, odim, odim) ⊗ (ptrace(A, [odim, idim], 1))^(-1/2)
+    A = B * A * B'
+    vA = represent(ChannelBasis{Matrix{ComplexF64}}(idim, odim), A)
+    Ap = combine(ChannelBasis{Matrix{ComplexF64}}(idim, odim), vA)
```
[the remark this answers is not in the document]
Comment (plewandowska777) [reply test/matrixbases.jl]: so, if you have suggestion about use constructor ChannelBasis, not the function channelbasis which we defined, you should also correct test 'represent, combine' :)

**Pull request merged**